src/libstd/thread/mod.rs

```diff
@@ -800,7 +800,11 @@ pub fn park() {
     match thread.inner.state.compare_exchange(EMPTY, PARKED, SeqCst, SeqCst) {
         Ok(_) => {}
         Err(NOTIFIED) => {
-            thread.inner.state.store(EMPTY, SeqCst);
+            // We must read again here, even though we know it will be NOTIFY,
+            // to synchronize with an write in `unpark` that occurred since we
+            // last read.
```

Comment:
To make sure I understand this, `NOTIFY` should be `NOTIFIED` and we're not actually synchronizing with anything in `unpark` but rather a write that happened *before* `unpark`?

Put another way lots of threads could be unparking this thread and wanting to send memory to it. We saw *some* threads via the `compare_exchange` above, but we're not guaranteed to see any other threads between that and the store of `EMPTY`. All the other threads in this small window are not notifying us, however. So we use `swap` here to make sure to maintain the invariant that we've synchronized with all threads who didn't notify us or the thread will notify us?

Reply:
Yes. So basically we want to establish acquire-release relationship with *every* `unpark()` that happened before this `park()`.

Reply:
@alexcrichton Yes, I will rewrite the comment to make that clearer.

Reply:
I guess one thing I'm not entirely sure of, doesn't `compare_exchange` already do the synchronization for us? When we come out the `Err` branch haven't we synchronized with all threads that already wrote `NOTIFIED` because we're for sure seeing `NOTIFIED`?

Reply:
`compare_exchange` does the synchronization for us, but it's not enough.

This is the crucial part you might be overlooking: In `unpark` now we unconditionally write `NOTIFIED` (using the `swap` operation) every time. Suppose that a thread calls `unpark` just after our `compare_exchange` was executed.

When we clear the notification signal (go from `NOTIFIED` to `EMPTY`), we want to synchronize with all threads that have delivered a notification so far. If we don't do a `swap` operation to clear notifications, we will not synchronize with the thread that called `unpark` just after our `compare_exchange`.

Reply:
Yes, I'll make the comment clearer like "since we last read NOTIFIED in the compare_exchange above".

This is quite interesting because it touches on some of the discussion in https://github.com/rust-lang/rfcs/pull/2503 in that, the semantics we really want here are store-acquire or a store-load barrier which aren't currently expressible in Rust (C11) atomics without doing a write.